src/service/resources-impl.js

```diff
@@ -407,13 +410,29 @@ export class Resources {
    * @private
    */
   buildReadyResources_() {
+    // Avoid cases where elements add more elements inside of them
+    // and cause an infinite loop of building - see #3354 for details.
+    if (this.isCurrentlyBuildingPendingResources_) {
```

Comment:
There might be other options to make sure of this, such as while-loop. Or we can have `buildReadyResourcesUnsafe_` call itself like this:
```
buildReadyResourcesUnsafe_() {
  let count = this.pendingBuildResources_.length;
  for (let i = 0; i < this.pendingBuildResources_.length; i++) {
    ...
    ...
    this.pendingBuildResources_.splice(i--, 1);
    count--;
    resource.build();
  }
  if (this.pendingBuildResources_.length > count) {
    this.buildReadyResourcesUnsafe_();
  }
}
```

Or maybe we can find another way (I don't really like recursion here). But I think it should be explicit since we explicitly ignoring build signal in a mutex.